### PR TITLE
ENH: Add support for tablewise application of style.background_gradient with axis=None

### LINKED ISSUE #15204: ENH: Allow elementwise coloring in background_gradient with axis=None
At present the axis argument to `background_gradient` only accepts 0 or 1 so that the resulting "heatmap" can only be applied columnwise or rowwise rather than to the whole table.
```
Signature: df.style.background_gradient(cmap='PuBu', low=0, high=0, axis=0, subset=None)

Parameters
----------
cmap: str or colormap
    matplotlib colormap
low, high: float
    compress the range by these values.
axis: int or str
    1 or 'columns' for colunwise, 0 or 'index' for rowwise
subset: IndexSlice
    a valid slice for ``data`` to limit the style application to
```
I propose that `axis=None` be allowed to mean applying a color gradient elementwise.

### PR DESCRIPTION
**Pull request by soxofaan**
- [x] closes #15204
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/formats/style.py [hunk 1]
```diff
@@ -901,16 +901,21 @@ def background_gradient(self, cmap='PuBu', low=0, high=0, axis=0,
     def _background_gradient(s, cmap='PuBu', low=0, high=0):
         """Color background in a range according to the data."""
         with _mpl(Styler.background_gradient) as (plt, colors):
-            rng = s.max() - s.min()
+            min = s.values.min()
+            max = s.values.max()
```
Comment (jschendel) [inline pandas/io/formats/style.py]: Can you use different names for `min`/`max` so they don't override the built-in `min`/`max` functions?

Comment (soxofaan) [reply pandas/io/formats/style.py]: sure, updated commit

**Comment by soxofaan** [comment]
@jschendel I updated the commit per your request

the failing tests/checks are not related to the change as far as I understand, is there something I should do about that?

**Comment by jreback** [comment]
you have linting issues: 

pandas/tests/io/formats/test_style.py:1066:1: E303 too many blank lines (3)

**Comment by jreback** [comment]
is it worth adding the OP example here?

**Comment by soxofaan** [comment]
@jreback   thanks, I didn't catch that linting issue warning. Lot of trees in the forest.

I think the OP example is good for interactive testing, but overkill for unit testing

**Comment by soxofaan** [comment]
(this PR will conflict with #21263 )

**Comment by soxofaan** [comment]
Had to rework this PR because #21263 has been merged

note that making `background_gradient` compatible with `axis=None` enabled to simplify the `relative_luminance` implementation of #21263

**Comment by soxofaan** [comment]
Illustration:
screen shot 2018-06-07 at 21 17 59

File: doc/source/whatsnew/v0.24.0.txt [hunk 2]
```diff
@@ -16,7 +16,9 @@ Other Enhancements
 - :func:`Series.mode` and :func:`DataFrame.mode` now support the ``dropna`` parameter which can be used to specify whether NaN/NaT values should be considered (:issue:`17534`)
 - :func:`to_csv` now supports ``compression`` keyword when a file handle is passed. (:issue:`21227`)
 - :meth:`Index.droplevel` is now implemented also for flat indexes, for compatibility with MultiIndex (:issue:`21115`)
-
+- Allow elementwise coloring in ``style.background_gradient`` with ``axis=None`` (:issue:`15204`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: use ``:class:`pandas.io.formats.style.Styler` ``

**Review by jreback — APPROVED** [review]
minor comments. lgtm.

File: doc/source/whatsnew/v0.24.0.txt (hunk 2 above)
Comment (soxofaan) [reply doc/source/whatsnew/v0.24.0.txt]: or even better, I'll move it alongside with the changelog entry for PR #21263

**Comment by soxofaan** [comment]
`continuous-integration/travis-ci/pr` failure is due to 
```
CondaHTTPError: HTTP 502 BAD GATEWAY for url <https://conda.anaconda.org/conda-forge/linux-64/repodata.json>
```

**Comment by soxofaan** [comment]
😕 now `continuous-integration/travis-ci/pr ` failure is due to 
```
$ curl -sSf -o python-3.5.tar.bz2 ${archive_url}
curl: (35) Unknown SSL protocol error in connection to s3.amazonaws.com:443 
Unable to download 3.5 archive. The archive may not exist. Please consider a different version.
```

**Review by TomAugspurger — APPROVED** [review]

**Comment by soxofaan** [comment]
(rebased)

**Comment by TomAugspurger** [comment]
Thanks!

**Pull request merged**